### PR TITLE
Automation on Bluestore pinned configuration parameter

### PR DESCRIPTION
**Pull request by SrinivasaBharath**
# Description

Automation of the polarion test case:
https://polarion.engineering.redhat.com/polarion/#/project/CEPH/workitem?id=CEPH-83575438
1. Configure cluster
2. Check the bluestore_cache_trim_max_skip_pinned variable value.
3. Create pools using the following command.
      ceph osd pool create <pool_name> <pgnum>
4. Push the data into the pool with rados bench.It depends on the size of the cluster.
    rados bench -p testbench  <timeinSec>write -b  <objectsize>--no-cleanup
5.  Start scrubbing and deep scrubbing on the cluster.
       command : ceph osd scrub all
6. Check for the crashes after finishing the scrub.

Following is the crash data:
  ()+0xf630) [0x7f2ae0162630]
  (BlueStore::Onode::decode(boost::intrusive_ptr<BlueStore::Collection>, ghobject_t const&, std::string const&, ceph::buffer::v14_2_0::list const&)+0x4c7) [0x565238a344d7]
 (BlueStore::Collection::get_onode(ghobject_t const&, bool)+0x856) [0x565238a6ab26]
 (BlueStore::stat(boost::intrusive_ptr<ObjectStore::CollectionImpl>&, ghobject_t const&, stat*, bool)+0x1a2) [0x565238a6b662]
 (PGBackend::be_scan_list(ScrubMap&, ScrubMapBuilder&)+0xe5) [0x5652388177b5]
 (PG::build_scrub_map_chunk(ScrubMap&, ScrubMapBuilder&, hobject_t, hobject_t, bool, ThreadPool::TPHandle&)+0x8b) [0x5652386c943b]
  (PG::chunky_scrub(ThreadPool::TPHandle&)+0x1a54) [0x5652386f3594]
 (PG::scrub(unsigned int, ThreadPool::TPHandle&)+0x4bb) [0x5652386f472b]
 (PGScrub::run(OSD*, OSDShard*, boost::intrusive_ptr<PG>&, ThreadPool::TPHandle&)+0x12) [0x56523889be42]
 (OSD::ShardedOpWQ::_process(unsigned int, ceph::heartbeat_handle_d*)+0x90f) [0x565238624bef]
 (ShardedThreadPool::shardedthreadpool_worker(unsigned int)+0x5b6) [0x565238be04d6]
 (ShardedThreadPool::WorkThreadSharded::entry()+0x10) [0x565238be2ff0]
 (()+0x7ea5) [0x7f2ae015aea5]
 (clone()+0x6d) [0x7f2adf01d98d]

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

**Comment by SrinivasaBharath** [comment]
Log file -  http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-3GKMK6

File: tests/rados/test_bluestore_features.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+            scrub_status = scrub_obj.verify_scrub_deepscrub(
+                before_scrub_log, after_scrub_log, "scrub"
+            )
+            deep_scrub_status = scrub_obj.verify_scrub_deepscrub(
+                before_deep_scrub_log, after_deep_scrub_log, "deepscrub"
+            )
+            log.info("scrubbing and deep-scrubbing are in progress")
+        log.info("scrubbing and deep-scrubbing completed")
+
+        # check fo r crashes in the cluster
```
Comment (pdhiran) [inline tests/rados/test_bluestore_features.py:75]: We need to continuously monitor the OSD memory growth behaviour at this point, when scrubbing is in progress.

The root cause for the test here is the memory growth, when it crosses a certain threshold, the OSD daemons would crash.

we need to adda method to check the memory usage of OSDs with scrubbing and IO's running. With ` bluestore_cache_trim_max_skip_pinned ` set to 1000 by default, we should see very less growth in OSD memory consumption.

Comment (pdhiran) [reply tests/rados/test_bluestore_features.py:75]: Ref : https://bugzilla.redhat.com/show_bug.cgi?id=1931504#c5

File: suites/quincy/rados/tier-2_rados_test_bluestore.yaml [hunk 2]
```diff
@@ -0,0 +1,127 @@
+# Suite contains basic tier-2 rados tests
```
Comment (pdhiran) [inline suites/quincy/rados/tier-2_rados_test_bluestore.yaml:1]: As discussed, please add the test to existing tier-2 test suites, instead creating a new one.

File: tests/rados/test_bluestore_features.py (hunk 1 above)
Comment (SrinivasaBharath) [reply tests/rados/test_bluestore_features.py]: I checking  the default  bluestore_cache_trim_max_skip_pinned value with 1000.Test case will fail if it is not equal to 1000.

**Comment by SrinivasaBharath** [comment]
Log File: http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-XQHTY6

File: tests/rados/test_bluestore_features.py (hunk 1 above)
Comment (neha-gangadhar) [reply tests/rados/test_bluestore_features.py]: Please add the check for OSD memory growth as well. getting the default value for 'bluestore_cache_trim_max_skip_pinned' and verifying seems very low ROI and not enough validation.

Comment (SrinivasaBharath) [reply tests/rados/test_bluestore_features.py]: Checking the memory growth while scrubbing.For every 10 seconds checking the total_used_bytes of the cluster. If the memory growth is more than 10% I am failing the test case.The code exist from line numbers 67 to 82.

File: tests/rados/test_bluestore_features.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    log.info(run.__doc__)
+    config = kw["config"]
+    cephadm = CephAdmin(cluster=ceph_cluster, **config)
+    rados_object = RadosOrchestrator(node=cephadm)
+    mon_obj = MonConfigMethods(rados_obj=rados_object)
+    scrub_obj = RadosScrubber(node=cephadm)
+    target_configs = config["cache_trim_max_skip_pinned"]["configurations"]
+
+    if config.get("cache_trim_max_skip_pinned"):
+        if not mon_obj.set_config(
```
[the remark this answers is not in the document]
Comment (SrinivasaBharath) [reply tests/rados/test_bluestore_features.py]: Checking the default bluestore_cache_trim_max_skip_pinned value with 1000.Test case will fail if it is not equal to 1000.

File: tests/rados/test_bluestore_features.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        # Creating pools and starting the test
+        for entry in target_configs.values():
+            log.debug(
+                f"Creating {entry['pool_type']} pool on the cluster with name {entry['pool_name']}"
+            )
+        method_should_succeed(
+            rados_object.create_pool,
+            **entry,
+        )
+        rados_object.bench_write(**entry)
```
[the remark this answers is not in the document]
Comment (SrinivasaBharath) [reply tests/rados/test_bluestore_features.py]: Added read and write  methods.

File: tests/rados/test_bluestore_features.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    max_skip_pinned_value = int(
+        mon_obj.get_config(section="osd", param="bluestore_cache_trim_max_skip_pinned")
+    )
+
+    # Check the default value of the  bluestore_cache_trim_max_skip_pinned value
+    if max_skip_pinned_value != 1000:
+        log.error(
+            "The default value of  bluestore_cache_trim_max_skip_pinned not equal to 1000"
+        )
+        return 1
```
Comment (pdhiran) [inline tests/rados/test_bluestore_features.py:41]: Please raise exceptions instead or return 1

File: tests/rados/test_bluestore_features.py (hunk 1 above)
Comment (pdhiran) [reply tests/rados/test_bluestore_features.py]: Hey Bharath, I think there is a confusion here. We are supposed to check the memory consumption grwoth of the OSD daemons, not the ceph cluster storage stats. 

Can you please check again? Similar activity is needed for task : RHCEPHQE-7917 . Ref : https://access.redhat.com/articles/351143

File: tests/rados/test_bluestore_features.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pdhiran) [reply tests/rados/test_bluestore_features.py]: Code changed to get config.

File: tests/rados/test_bluestore_features.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    log.info("Finished writing data into the pool")
+    rados_object.bench_write(**entry)
+
+    # performing scrub and deep-scrub
+    rados_object.run_scrub()
+    rados_object.run_deep_scrub()
+    time.sleep(10)
+
+    # Checking for the memory growth
+    cluster_stats = get_storage_stats(client_node)
```
Comment (pdhiran) [inline tests/rados/test_bluestore_features.py:66]: Hey Bharath, I think there is a confusion here. We are supposed to check the memory consumption growth of the OSD daemons, not the ceph cluster storage stats.

Can you please check again? Similar activity is needed for task : RHCEPHQE-7917 . Ref : https://access.redhat.com/articles/351143

**Review by pdhiran — CHANGES_REQUESTED** [review]
Code needs to be updated from checking ceph storage stats to checking OSD daemon memory utilisation checks.

File: tests/rados/test_bluestore_features.py (hunk 5 above)
Comment (SrinivasaBharath) [reply tests/rados/test_bluestore_features.py]: Raised exception.

File: tests/rados/test_bluestore_features.py (hunk 1 above)
Comment (SrinivasaBharath) [reply tests/rados/test_bluestore_features.py]: Calculating the memory used by every OSD by using the heap dump command. The sample output of heap dump output is as below-
_-----------------------------------------------
MALLOC:      538211456 (  513.3 MiB) Bytes in use by application
MALLOC: +            0 (    0.0 MiB) Bytes in page heap freelist
MALLOC: +     14341008 (   13.7 MiB) Bytes in central cache freelist
MALLOC: +      4471808 (    4.3 MiB) Bytes in transfer cache freelist
MALLOC: +     19725296 (   18.8 MiB) Bytes in thread cache freelists
MALLOC: +      3670016 (    3.5 MiB) Bytes in malloc metadata
MALLOC:   ------------
MALLOC: =    580419584 (  553.5 MiB) Actual memory used (physical + swap)
MALLOC: +    197885952 (  188.7 MiB) Bytes released to OS (aka unmapped)
MALLOC:   ------------
MALLOC: =    778305536 (  742.2 MiB) Virtual address space used
MALLOC:
MALLOC:          18491              Spans in use
MALLOC:             34              Thread heaps in use
MALLOC:           8192              Tcmalloc page size
------------------------------------------------_ 

Executing the tests for 45 minutes and every 10 seconds the details are storing in a list.Once the data is retrieved the relative growth is calculating for every OSD.If the growth is greater than 80% then making testcase fail. 

To get the 80% benchmark , I executed the tests with the 256KB and 64MB.For 256KB the memory growth is very minimal so I choose the 64MB.In addition to that, executed the tests with the low "bluestore_cache_trim_max_skip_pinned" value that is 10 and noticed that the memory growth some time more than 100% but with the default value that growth is in between 60-80%.

**Memory growth when bluestore_cache_trim_max_skip_pinned=10 & objectsize=64MB:**
2023-04-08 19:21:16,853 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.0 consuming more memory with the memory growth 44.58713547406898
2023-04-08 19:21:16,854 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.0 is 44.58713547406898 
2023-04-08 19:21:16,856 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.3 consuming more memory with the memory growth 145.84464591690133
2023-04-08 19:21:16,857 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.3 is 145.84464591690133 
2023-04-08 19:21:16,858 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.6 consuming more memory with the memory growth 41.93759008922337
2023-04-08 19:21:16,859 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.6 is 41.93759008922337 
2023-04-08 19:21:16,860 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.9 consuming more memory with the memory growth 53.49275179438129
2023-04-08 19:21:16,861 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.9 is 53.49275179438129 
2023-04-08 19:21:16,862 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.12 consuming more memory with the memory growth 45.60211673729597
2023-04-08 19:21:16,863 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.12 is 45.60211673729597 
2023-04-08 19:21:16,864 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.2 consuming more memory with the memory growth 38.50719602435498
2023-04-08 19:21:16,864 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.2 is 38.50719602435498 
2023-04-08 19:21:16,865 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.5 consuming more memory with the memory growth 68.27855738178029
2023-04-08 19:21:16,866 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.5 is 68.27855738178029 
2023-04-08 19:21:16,866 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.8 consuming more memory with the memory growth 32.06705562117973
2023-04-08 19:21:16,867 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.8 is 32.06705562117973 
2023-04-08 19:21:16,868 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.11 consuming more memory with the memory growth 45.774732483609924
2023-04-08 19:21:16,868 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.11 is 45.774732483609924 
2023-04-08 19:21:16,869 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.14 consuming more memory with the memory growth 30.69973365755619
2023-04-08 19:21:16,869 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.14 is 30.69973365755619 
2023-04-08 19:21:16,870 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.1 consuming more memory with the memory growth 26.564678005489018
2023-04-08 19:21:16,871 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.1 is 26.564678005489018 
2023-04-08 19:21:16,871 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.4 consuming more memory with the memory growth 68.50086699397161
2023-04-08 19:21:16,872 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.4 is 68.50086699397161 
2023-04-08 19:21:16,873 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.7 consuming more memory with the memory growth 55.005894388812855
2023-04-08 19:21:16,873 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.7 is 55.005894388812855 
2023-04-08 19:21:16,874 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.10 consuming more memory with the memory growth 49.51864084399263
2023-04-08 19:21:16,875 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.10 is 49.51864084399263 
2023-04-08 19:21:16,875 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.13 consuming more memory with the memory growth 54.60954384746342
2023-04-08 19:21:16,876 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.13 is 54.60954384746342

Log files:
   1.http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-WQTRYO
   2.http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-JVYXEL

File: tests/rados/test_bluestore_features.py (hunk 6 above)
Comment (SrinivasaBharath) [reply tests/rados/test_bluestore_features.py]: Code modified and explained above.

File: suites/quincy/rados/tier-2_rados_test_bluestore.yaml [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+  - test:
+      name:  Testing bluestore
+      desc: Testing bluestore pinned
+      module: test_bluestore_features.py
+      polarion-id: CEPH-83575438
+      config:
+          cache_trim_max_skip_pinned:
+            configurations:
+               pool-1:
+                 pool_type: replicated
```
Comment (neha-gangadhar) [inline suites/quincy/rados/tier-2_rados_test_bluestore.yaml:117]: Can we also verify on ec pool ?

File: suites/quincy/rados/tier-2_rados_test_bluestore.yaml [hunk 8]
```diff
@@ -0,0 +1,125 @@
+# Suite contains basic tier-2 rados tests
```
Comment (neha-gangadhar) [inline suites/quincy/rados/tier-2_rados_test_bluestore.yaml:1]: Please add the suite for pacific as well.

File: suites/quincy/rados/tier-2_rados_test_bluestore.yaml (hunk 7 above)
Comment (SrinivasaBharath) [reply suites/quincy/rados/tier-2_rados_test_bluestore.yaml:117]: Add the EC pool tests and code.

**Comment by SrinivasaBharath** [comment]
Log file : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-93CW8Z

File: suites/quincy/rados/tier-2_rados_test_bluestore.yaml (hunk 8 above)
Comment (SrinivasaBharath) [reply suites/quincy/rados/tier-2_rados_test_bluestore.yaml:1]: Added the pacific suite

File: ceph/rados/core_workflows.py [hunk 9]
```diff
@@ -574,6 +575,28 @@ def enable_balancer(self, **kwargs) -> bool:
         log.info(f"the balancer status is \n {out}")
         return True
 
+    def check_file_exists_on_client(self, loc) -> bool:
```
Comment (pdhiran) [inline ceph/rados/core_workflows.py:578]: This method was already added as part of this PR : https://github.com/red-hat-storage/cephci/pull/2426/files and merged.

Please do a clean rebase of the code and push the changes.

Comment (SrinivasaBharath) [reply ceph/rados/core_workflows.py]: Please check, did the rebase and pushed the code.

File: tests/rados/test_bluestore_features.py [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+        malloc_str = osd_dump.split("MALLOC:")
+        result = [app_dump for app_dump in malloc_str if app_str in app_dump]
+        result = (" ".join(result)).strip()
+        result_list = result.split("(")
+        osd_id_app_mem[osd_id] = result_list[0].strip()
+    return osd_id_app_mem
+
+
+# Merge dictionaries
+def mergeDictionary(dict_1, dict_2):
```
Comment (pdhiran) [inline tests/rados/test_bluestore_features.py:151]: can we use built-in method for this? dict1.update(dict2) ? The return would be exactly same

**Review by pdhiran — COMMENTED** [review]
Looks good with minor changes needed.

File: tests/rados/test_bluestore_features.py (hunk 10 above)
Comment (SrinivasaBharath) [reply tests/rados/test_bluestore_features.py:151]: With the update, the new value of the key will get replaced with the old value. Here I am appending the new memory value to the key.

Comment (pdhiran) [reply tests/rados/test_bluestore_features.py:151]: Please update the docstrings appropriately, adding summary, usahe, examples and return types with examples.

File: ceph/rados/core_workflows.py [hunk 11]
```diff
@@ -1407,3 +1407,37 @@ def get_osd_hosts(self):
...
+        if not osd_list:
+            log.error("OSD list is empty")
+            return 1
+        for osd_id in osd_list:
+            cmd = f"ceph tell osd.{osd_id} heap {action}_profiler"
+            self.node.shell([cmd])
+        log.info(f"The OSD {osd_list} heap profile is in {action} state")
+        return 0
+
+    def get_heap_dump(self, osd_list):
```
Comment (pdhiran) [inline ceph/rados/core_workflows.py:1428]: Please update the docstrings appropriately, adding summary, usahe, examples and return types with examples.

Comment (SrinivasaBharath) [reply ceph/rados/core_workflows.py:1430]: Added the doc strings.

File: tests/rados/test_bluestore_features.py (hunk 10 above)
Comment (SrinivasaBharath) [reply tests/rados/test_bluestore_features.py:151]: Doc string is added

**Review by pdhiran — COMMENTED** [review]
Looks good!

**Review by pdhiran — APPROVED** [review]

**Pull request merged**